pkg/lorawan/status.go

```diff
@@ -0,0 +1,161 @@
+// Package lorawan holds decoding functions for lorawan (Macaron) status payloads.
+package lorawan
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/DIMO-Network/model-garage/pkg/cloudevent"
+	"github.com/DIMO-Network/model-garage/pkg/convert"
+	"github.com/segmentio/ksuid"
+	"github.com/tidwall/gjson"
+)
+
+const (
+	FingerprintFrame = 0x01
+	SleepFrame       = 0x05
+	DataVersion      = "v2"
+)
+
+// TokenIDFromData gets a tokenID from a V2 payload.
+func TokenIDFromData(jsonData []byte) (uint32, error) {
+	lookupKey := "subject"
+	subject := gjson.GetBytes(jsonData, lookupKey)
+	if !subject.Exists() {
+		return 0, convert.FieldNotFoundError{Field: "tokenID", Lookup: lookupKey}
+	}
+	subjectStr, ok := subject.Value().(string)
+	if !ok {
+		return 0, fmt.Errorf("%s field is not a string", lookupKey)
+	}
+	subjectDID, err := cloudevent.DecodeNFTDID(subjectStr)
+	if err != nil {
+		return 0, fmt.Errorf("error decoding subject: %w", err)
+	}
+	return subjectDID.TokenID, nil
+}
+
+// SourceFromData gets a source from a V2 payload.
+func SourceFromData(jsonData []byte) (string, error) {
+	lookupKey := "source"
+	source := gjson.GetBytes(jsonData, lookupKey)
+	if !source.Exists() {
+		return "", convert.FieldNotFoundError{Field: "source", Lookup: lookupKey}
+	}
+	src, ok := source.Value().(string)
+	if !ok {
+		return "", errors.New("source field is not a string")
+	}
+	return src, nil
+}
+
+// ConvertToCloudEvents converts a message data payload into a slice of CloudEvents.
+// It handles both status and fingerprint events, creating separate CloudEvents for each.
+func ConvertToCloudEvents(msgData []byte, chainID uint64, aftermarketContractAddr, vehicleContractAddr string) ([][]byte, error) {
```

Comment:
I don't think we should return a list of cloudevent objects. Dis only wants a list of headers and a single body
https://github.com/DIMO-Network/dis/blob/898763c0c7f0dd2383da5df6879066d0790568ae/internal/modules/macaron/macaron.go#L87-L103
```suggestion
func ConvertToCloudEvents(msgData []byte, chainID uint64, aftermarketContractAddr, vehicleContractAddr string) ([]cloudevent.CloudEventHeader, []byte, error){
```

Reply:
I've used the same convention as for autopi, but if all those events share the same data, then I agree that we can simplify this part to simplify DIS code.

On the other hand creating separate cloud events here makes it more generic as a public reusable code.

Reply:
For this one I think AP should be updated as well, there isn't a lot of public usage, but I am not going to block the PR over it.